feat: force enrollment after enrollment end date

aliraza-abbasi opened the pull request:
<!--

🫒🫒
🫒🫒🫒🫒         🫒 Note: the Olive master branch has been created.  Please consider whether your change
    🫒🫒🫒🫒     should also be applied to Olive. If so, make another pull request against the
🫒🫒🫒🫒         open-release/olive.master branch, or ping @nedbat for help or questions.
🫒🫒

🌰🌰🌰🌰🌰🌰     🌰 Note: the Nutmeg release is still supported.
                  Please consider whether your change should be applied to Nutmeg as well.

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/openedx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

It includes support for course enrollment in case of `enrollment_end` date has passed or the `upgrade_deadline` has passed. The `force_enrollment` argument is used to support this functionality, and `can_upgrade` and `include_expired` will be `True` if `force_enrollment` is `True`. Only a user who has `GlobalSupport` access can perform this operation. 
This feature is required because as an admin or support user I need to create verified enrollments sometimes:

- to address customer support issues
- to test the system
- to enroll staff

These verified enrollments need to be accounted for properly, so they should be created with an enrollment code.

Sometimes I need to do this after the `enrollment_end` date has passed, or the `upgrade_deadline` has passed.

## Related Issue
https://github.com/openedx/edx-platform/issues/31308

openedx-webhooks commented:
Thanks for the pull request, @aliraza-abbasi!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pdpinch commented:
It looks like you need a rebase.

aliraza-abbasi commented:
> It looks like you need a rebase.

I've rebased it multiple times the new changes are being merged very quickly. I'll rebase it after the review

asadali145 reviewed (commented):
LGTM! Could we please add tests for forced enrollments?

pdpinch commented:
Does this work to create a verified enrollment after the upgrade deadline has passed?

aliraza-abbasi commented:
> Does this work to create a verified enrollment after the upgrade deadline has passed? 

Yes, it worked. I've tested the API

asadali145 commented on openedx/core/djangoapps/enrollments/api.py:
```diff
@@ -199,7 +199,16 @@ def get_enrollment(username, course_id):
+    enrollment_attributes=None,
+    enterprise_uuid=None,
+    force_enrollment=False,
+    include_expired=False
```
Can we add `include_expired ` in docstring args?

aliraza-abbasi replied:
fixed

asadali145 reviewed (approved):
Looks good to me!

pdpinch commented:
@aliraza-abbasi can you squash your commits and write a more descriptive commit message? Something similar to the PR description would be fine:

```
feat: force enrollment after enrollment end date via API

It includes support for course enrollment in case of enrollment_end date has passed or the upgrade_deadline has passed. The force_enrollment argument is used to support this functionality, and check_access will be False if force_enrollment is True. Only a user who has GlobalSupport access can perform this operation.
```

aliraza-abbasi committed:
feat: force enrollment after enrollment end date via API

    It includes support for course enrollment in case of enrollment_end date has passed or the upgrade_deadline has passed. The force_enrollment argument is used to support this functionality, and can_upgrade and include_expired will be True if force_enrollment is True. Only a user who has GlobalSupport access can perform this operation.

openedx-webhooks commented:
@aliraza-abbasi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31309&entry.1484655318=https://github.com/aliraza-abbasi&entry.752974735=2022-11-18+14:09&entry.1917517419=2022-12-08+20:58&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
